frontend/src/use-url-state.ts

```diff
@@ -0,0 +1,85 @@
+import { useState, useCallback, useEffect } from 'react';
+
+type state = string|null;
+type stateSetter = (state: state) => void;
+type stringSetter = (state: string) => void;
+
+type useUrlStateReturn = [
+   string,
+   stateSetter,
+];
+
+type HistoryState = Record<string, string>;
+
+/**
+ * like useState(), but the state is stored in the url as a query param.
+ * Loads it from a query param on page refresh and updates the url as the state
+ * changes.
+ * paramName is expected not to change over the life of the component.
+ */
+export function useUrlState(paramName: string, paramDefault: string): useUrlStateReturn  {
+   const [state, setState] = useState(paramDefault);
+
+   // Load state from url, only once per component
+   useEffect(() => {
+      watchForPopstate(paramName, setState, paramDefault);
```

Comment:
I don't know if it's important, but you might consider making this return a function to remove the event listener, and returning that function from the `useEffect`. It will then be run "before the component leaves the screen" according to the [docs](https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect), which allows us to stop getting events that try to set state on a removed React component.

Reply:
Good point. I didn't realize that happens, but of course it does.

Reply:
Fixed!